src/Message/MessageFactory.php

```diff
@@ -8,12 +8,77 @@
 use RingCentral\Psr7\Response;
 use RingCentral\Psr7\Uri;
 use React\Stream\ReadableStreamInterface;
+use Clue\React\Zlib\ZlibFilterStream;
 
 /**
  * @internal
  */
 class MessageFactory
 {
+    /**
+     * Creates an associative array of lowercase header names to the actual header
+     *  casing.
+     *
+     * @param array $headers Associative array of HTML headers
+     *
+     * @return array
+     */
+    private static function _normalizeHeaderKeys(array $headers)
+    {
+        $result = array();
+
+        foreach (array_keys($headers) as $key) {
+            $result[strtolower($key)] = $key;
+        }
+
+        return $result;
+    }
+
+    /**
+     * Automatically decode (decompress) responses when instructed.
+     *
+     * @param array                          $options Response options
+     * @param array                          $headers Response headers
+     * @param ReadableStreamInterface|string $body    Response body
+     * 
+     * @return ReadableStreamInterface|string
+     */
+    private static function _checkDecode(array $options, array &$headers, $body)
+    {
+        if (!empty($options['decodeContent'])
+            && $body instanceof ReadableStreamInterface
+            && extension_loaded('zlib')
+        ) {
+            $normalizedKeys = self::_normalizeHeaderKeys($headers);
+
+            if (isset($normalizedKeys['content-encoding'])) {
+                $encoding = $headers[$normalizedKeys['content-encoding']];
+
+                if ($encoding === 'gzip' || $encoding === 'deflate') {
+                    $body = $body->pipe(
+                        $encoding === 'gzip'
+                        ? ZlibFilterStream::createGzipDecompressor()
+                        : ZlibFilterStream::createZlibDecompressor()
+                    );
+
+                    // Remove content-encoding header
+                    $headers['x-encoded-content-encoding']
+                        = $headers[$normalizedKeys['content-encoding']];
+                    unset($headers[$normalizedKeys['content-encoding']]);
```

Comment:
I understand the motivation for this change, but do we really need to remove these headers? It's my understanding that if this behavior has to be turned on explicitly, most consumers shouldn't care about these headers anyway and they mostly serve an informational purpose. This would also be in line with e.g. curl's `--compressed` option?

Reply:
These changes were based on the solution used by Guzzle, PHP HTTP client, and they do make sense to me. However, if you do not want to include them, it would not hurt me.